object avoidance project + I2C mutex + updated docs

RobertLucian opened the pull request:
In this PR, I'm bringing a project built on top of the `GoPiGo3`. 

It's a small example project, where the robot is supposed to move forward as long as it's not too close to a target. Its speed is proportional to the distance from the target.

The following issues have also been fixed:

* https://github.com/RobertLucian/GoPiGo3/issues/10
* https://github.com/RobertLucian/GoPiGo3/issues/9
* https://github.com/RobertLucian/GoPiGo3/issues/7


----------

Along with this project, I've also modified the way the mutex on `I2C` works. Instead of using a function, there's now a class from which another class can derive. Also, the user can now set whether `DistanceSensor` and `DHTSensor` objects are multithreaded.

-----------

Here's a link to the [project](https://github.com/RobertLucian/GoPiGo3/tree/feature/object-avoidance-project/Projects/ObjectAvoidanceRobot). The link will be valid as long as the PR is open.

-----------

Before accepting this PR, it's also needed to bring the required modifications to the [GoPiGo2](https://github.com/DexterInd/GoPiGo) repository.

-----------

The documentation has been updated with the latest changes and it looks like this:

![distance](https://user-images.githubusercontent.com/26958764/29143544-46fb8938-7d5e-11e7-9dff-3db1dc7e2dc6.PNG)

RobertLucian commented:
After a couple of small mods, I've tested both the `DistanceSensor` and `DHTSensor` classes and they worked as expected both when using the `mutex` and when not.

johnisanerd commented on Software/Python/easygopigo3.py:
```diff
@@ -28,8 +28,6 @@
 except:
```
This is the second one I'm worried about.  I will defer to @CleoQc  on this one though.  We might also have @mattallen37  look at it.

johnisanerd commented on Software/Python/I2C_mutex.py:
```diff
@@ -1,27 +1,56 @@
 import fcntl
```
This is the First one I'm worried about.  I will defer to @CleoQc  on this one though.  We might also have @mattallen37  look at it.

RobertLucian replied:
Why are you afraid of the `fcntl`? What are you thinking of?

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2453,13 +2440,16 @@ def read_mm(self):
         # if sensor insists on that value, then pass it on
+
         while (mm > 8000 or mm < 5) and attempt < 3:
+            if self.mutex:
```
Put these and all others like it into one line

RobertLucian replied:
It's done and tested.

RobertLucian commented:
I've solved the conflicts in this one. Apparently, the latest changes made this PR un-merge-able.

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2756,13 +2798,9 @@ def read_humidity(self):
-        finally:
-            _release_read()
+        self.__ifMutexAcquire()
+        humidity = DHT.dht(self.sensor_type)[1]
```
This should be within Try/Except/Finally

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2733,13 +2779,9 @@ def read_temperature(self):
-        finally:
-            _release_read()
+        self.__ifMutexAcquire()
+        temp = DHT.dht(self.sensor_type)[0]
```
Try/Except/Finally

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2644,14 +2650,12 @@ def read_mm(self):
-            finally:
-                _release_read()
-                
+            self.__ifMutexRelease()
```
To be kind to our future self, the mutexRelease is under finally:

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2724,6 +2728,48 @@ def __init__(self, port="SERIAL",gpg=None, sensor_type=0):
             raise
 
+        try:
+            import threading
```
is that new code?

RobertLucian replied:
Really good catch. This an old remnant.

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2775,13 +2813,9 @@ def read_humidity(self):
-            raise
-        finally:
-            _release_read()
+        self.__ifMutexAcquire()
```
try/except/finally

RobertLucian commented:
@CleoQc so I've gone through the source code and I've noticed really peculiar things - nevertheless I've fixed them (probably because of the former conflicts things got messed up), added support for the DHT sensor in the documentation, polished a bit the DHT sensor and made some further changes to the documentation.

There's one last thing. The thing with the mutexes has been done the following way:
```
acquireMutex
try:
    do whatever
except:
    throw whatever
releaseMutex
```
How does this sound to you? My motive was that I didn't want to do the acquiring and releasing of mutexes inside the `try/except/finally` blocks. Even if the `finally` block gets executed no matter what, it feels and looks better to keep it out of it. What do you think? Do you have a specific reason to keep in there? I saw you said *to be kind to our future self* and I wonder what you're thinking about.

RobertLucian commented:
@CleoQc I've tested it and I've added the requested changes so it's all good for a review now.

CleoQc commented on Software/Python/I2C_mutex.py:
```diff
@@ -1,27 +1,56 @@
-        except IOError: # already locked by a different process
+            DexterLockI2C_handle = open('/run/lock/DexterLockI2C')
+        except:
+            pass
```
what may cause this to fail? Wouldn't it be best to at least print the exception?

RobertLucian replied:
No. There are 2 reasons:

1. Doing it so will lead to a very ugly console output, provided someone does use multiple threads/processes. Basically, the console will be filled with the said message.
2. Second of all, this is considered to be a bad practice. If we would want to do something about it, we could create a timeout on this method (most probably through another specialized class) and then end the method when the time runs out. Another option is using modules that are intended to be used for logging the output. Bottom line is: we should steer away from prints.

Also, this fails whenever the resource is locked by another thread/process - I wrote there a comment.

CleoQc replied:
Ok, you're right. All good then on this point

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2724,71 +2750,81 @@ def __init__(self, port="SERIAL",gpg=None, sensor_type=0):
-            raise
-        finally:
-            _release_read()
+        _ifMutexAcquire(self.use_mutex)
```
this used to be in a try/except. Even if dht sensor doesn't actually raise an error, I think it's best to keep everything the same.

RobertLucian replied:
You mean about keeping the `ifMutexWhatever` inside the finally block? If so, I already did that, but it doesn't show up.

CleoQc commented on Software/Python/easygopigo3.py:
```diff
@@ -2724,71 +2750,81 @@ def __init__(self, port="SERIAL",gpg=None, sensor_type=0):
-        finally:
-            _release_read()
+        _ifMutexAcquire(self.use_mutex)
+        [temp , humidity]=DHT.dht(self.sensor_type)
```
Same here. Please keep the try/except/finally structure
